### PR TITLE
Plans 2023: Adjust the select Plans functionality in the compare grid, for the /plans page

### PR DESCRIPTION
**Pull request by claudiucelfilip**
#### Proposed Changes

This adjusts the Plans comparison grid with the following changes;
* handles alignment with missing CTA buttons
* add current plan CTA styling
* show the Free plan if it's the current plan
* reduces the number of columns and adds the Plan selection functionality. In this situation, the current plan is always first.

NOTE: There's a problem at around 880px where the first plan overlaps the Feature labels. Normally, we'd switch to mobile view (which hides the Feature labels) at this breakpoint but having a 272px sidebar means we'd need the switch to happen at 880 + 272px. Will need to follow up with a fix.

#### Testing Instructions

* Go to `/plans/{CURRENT_SITE}?flags=onboarding/2023-pricing-grid` - where CURRENT_SITE is a Free plan
* 
* You should see the `Free` plan in the Plans and Plan Comparison Grids
* Go through different viewport sizes; the column number should adjust so that they can fit
* When there are fewer plans visible than the total number, the current plan should be first and un-selectable. All other plans should be selectable

https://user-images.githubusercontent.com/2749938/215505077-78c9837c-2491-44b0-ab84-2d56c3cd3c76.mp4

* Switch to a `Business` plan (or any paid plan)
* Free plan should not be visible anymore
* Make sure the current plan is **visible** on all viewport sizes

https://user-images.githubusercontent.com/2749938/215505232-b206c616-5828-445a-bd2c-d047f958f182.mp4

* Go to `/start/onboarding-2023-pricing-grid/plans?flags=onboarding/2023-pricing-grid`
* The page should be **unaffected**

#### Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

Related to https://github.com/Automattic/martech/issues/1448 and https://github.com/Automattic/martech/issues/1446

### CODE REVIEW

File: client/my-sites/plans/main.jsx [hunk 1]
```diff
@@ -143,11 +143,14 @@ class Plans extends Component {
 			);
 		}
 
+		const hideFreePlan =
+			! isEnabled( 'onboarding/2023-pricing-grid' ) || currentPlan.productSlug !== 'free_plan';
```
Comment (claudiucelfilip) [inline client/my-sites/plans/main.jsx]: @chriskmnds This might overlap with your work on https://github.com/Automattic/martech/issues/1450 - this allows the Free plan to reach the new Plans and Plan Comparison Grids, if it's the current plan.

File: client/my-sites/plan-features-2023-grid/style.scss [hunk 2]
```diff
@@ -875,6 +885,9 @@ body.is-section-signup.is-white-signup .is-onboarding-2023-pricing-grid .signup_
 }
 
 .plan-comparison-grid {
+	.plan-features-2023-gridrison__actions-buttons {
+		min-height: 40px;
```
Comment (claudiucelfilip) [inline client/my-sites/plan-features-2023-grid/style.scss:889]: Ensures buttons are aligned.

File: client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx [hunk 3]
```diff
@@ -373,16 +377,31 @@ export const PlanComparisonGrid: React.FC< PlanComparisonGridProps > = ( {
...
 		}
 
 		setVisiblePlans( newVisiblePlans );
-	}, [ isMediumBreakpoint, intervalType ] );
+	}, [
+		isLargestBreakpoint,
+		isLargeBreakpoint,
+		isMediumBreakpoint,
+		intervalType,
+		planProperties,
```
Comment (claudiucelfilip) [inline client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx]: We need to add `planProperties` to ensure that switching between sites (Free vs Paid) updates the plan list.
Using `displayedPlansProperties`  seems to create an infinite loop.

File: client/my-sites/plans/main.jsx (hunk 1 above)
Comment (chriskmnds) [reply client/my-sites/plans/main.jsx]: Thanks for mentioning @claudiucelfilip . Indeed.

I guess we are bound to reach these overlaps if some of us are working on general open-ended tickets like the one you linked to from the description.

> this allows the Free plan to reach the new Plans and Plan Comparison Grids, **if it's the current plan**

It's not visible to me where this has been decided. I've been trying to figure this out in Slack today. I understand this is part of design/intention? 🤔

Comment (claudiucelfilip) [reply client/my-sites/plans/main.jsx]: The idea is to hide the Free plan if you're on a Paid plan. I kept it simple and just checked for the free plan, but for a complete approach, we might want to use a selector to properly check if the current plan is paid or not.

File: client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx [hunk 4]
```diff
@@ -361,8 +364,9 @@ export const PlanComparisonGrid: React.FC< PlanComparisonGridProps > = ( {
 		( { planName } ) => ! ( planName === PLAN_ENTERPRISE_GRID_WPCOM )
 	);
 	const isMonthly = intervalType === 'monthly';
+	const isLargestBreakpoint = usePricingBreakpoint( 1600 );
+	const isLargeBreakpoint = usePricingBreakpoint( 1500 );
```
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx]: Where do we get these values from (1600, 1500)? My thinking is we should be aligned with some more universal list of breakpoints, or adding to it, or at least naming and exposing these more centrally.

File: client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx [hunk 5]
```diff
@@ -373,16 +377,31 @@ export const PlanComparisonGrid: React.FC< PlanComparisonGridProps > = ( {
...
-		if ( isSmallBreakpoint || ( isInSignup && isMediumBreakpoint ) ) {
-			newVisiblePlans = newVisiblePlans.slice( 0, 2 );
+		visibleLength = isMediumBreakpoint ? 2 : visibleLength;
+
+		if ( newVisiblePlans.length !== visibleLength ) {
+			const currentPlan = displayedPlansProperties.find( ( { current } ) => current );
+			if ( currentPlan ) {
+				newVisiblePlans.sort( ( a ) => ( a === currentPlan.planName ? -1 : 1 ) );
+			}
+			newVisiblePlans = newVisiblePlans.slice( 0, visibleLength );
```
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx]: This chunk is hard to follow. I wish we could simplify a bit or contextualize a little better. Let's see if the final shape is a little cleaner - otherwise, maybe avoiding the reassignments might force us to `name` an extra variable or two and speak more clearly. WDYT?

File: client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx [hunk 6]
```diff
@@ -373,16 +377,31 @@ export const PlanComparisonGrid: React.FC< PlanComparisonGridProps > = ( {
...
+			visibleLength = isLargestBreakpoint ? 4 : visibleLength;
+			visibleLength = isLargeBreakpoint ? 3 : visibleLength;
 		}
 
-		if ( isSmallBreakpoint || ( isInSignup && isMediumBreakpoint ) ) {
-			newVisiblePlans = newVisiblePlans.slice( 0, 2 );
+		visibleLength = isMediumBreakpoint ? 2 : visibleLength;
+
+		if ( newVisiblePlans.length !== visibleLength ) {
+			const currentPlan = displayedPlansProperties.find( ( { current } ) => current );
```
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx]: Is "current plan" not visible from props i.e. can we avoid looping here?

File: client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx (hunk 3 above)
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx]: If we do indeed need `displayedPlansProperties` then the infinite loop is a signal we may be doing something wrong. Probably worth investigating?

File: client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx [hunk 7]
```diff
@@ -500,9 +519,8 @@ export const PlanComparisonGrid: React.FC< PlanComparisonGridProps > = ( {
 					const featureGroupClass = `feature-group-title-${ featureGroup.slug }`;
 					const isHiddenInMobile = ! visibleFeatureGroups.includes( featureGroup.slug );
 					return (
-						<div className={ featureGroupClass }>
+						<div key={ featureGroupClass } className={ featureGroupClass }>
```
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx]: NIT - why not `featureGroup.slug` for the key? For clarity's sake, the "title" sits below this element.

File: client/my-sites/plan-features-2023-grid/style.scss [hunk 8]
```diff
@@ -875,6 +885,9 @@ body.is-section-signup.is-white-signup .is-onboarding-2023-pricing-grid .signup_
 }
 
 .plan-comparison-grid {
+	.plan-features-2023-gridrison__actions-buttons {
```
Comment (chriskmnds) [inline client/my-sites/plan-features-2023-grid/style.scss:888]: What does `gridrison` mean? Is this an acronym for "grid-comparison"? 😁 
Would a less fictitious name work here, maybe `plan-features-2023-grid-actions__actions-buttons`?

File: client/my-sites/plans/main.jsx [hunk 9]
```diff
@@ -143,11 +144,14 @@ class Plans extends Component {
 			);
 		}
 
+		const hideFreePlan =
+			! isEnabled( 'onboarding/2023-pricing-grid' ) || ! isFreePlan( currentPlan.productSlug );
```
Comment (chriskmnds) [inline client/my-sites/plans/main.jsx]:
```suggestion
const hideFreePlan = isEnabled('onboarding/2023-pricing-grid') 
    ? ! isFreePlan(currentPlan.productSlug) 
    : true;
```
I find the logical OR with negations on both sides a little hard to follow. The ternary form feels a little more concise and easier to read. The two should be identical?

**Review by chriskmnds — COMMENTED** [review]
Hey @claudiucelfilip. I left some code comments/suggestions below. If nothing critical among those, then I'll take it for a spin tomorrow and approve.

File: client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx (hunk 3 above)
Comment (ddc22) [reply client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx]: @chriskmnds is correct. This happens because of the code here.
https://github.com/Automattic/wp-calypso/blob/24a9d72f43ba6ba0a4fa32e756be040b3ddde31f/client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx#L363-L365

The filter continously creates a new array reference on each render, which triggers a call to this useEffect when displayedPlansProperties is in the dependency array. The use effect also essentially set's a new array as state every-time it is triggered. This is what creates the infinite loop.

Wrapping the [above line](https://github.com/Automattic/wp-calypso/blob/24a9d72f43ba6ba0a4fa32e756be040b3ddde31f/client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx#L363-L365) with [useMemo](https://beta.reactjs.org/reference/react/useMemo) should fix it.

Comment (claudiucelfilip) [reply client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx]: Good points and thanks for the suggestions!
useMemo seems to fix it.

File: client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx (hunk 4 above)
Comment (claudiucelfilip) [reply client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx]: The base breakpoints can be found in this [SASS file](https://github.com/Automattic/wp-calypso/blob/trunk/client/signup/steps/plans/style.scss#L1).
Considering we have a lefthand sidebar, we'll need another set of these for /plans, and there's a [task for it](https://github.com/Automattic/martech/issues/1455).

For the scope of this PR, I adjusted these to be more conservative (higher) and matched the original breakpoint + 272px (width of the sidebar).

File: client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx (hunk 5 above)
Comment (claudiucelfilip) [reply client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx]: > otherwise, maybe avoiding the reassignments might force us to name an extra variable or two and speak more clearly.

I'm not sure what you mean, but I agree this can be improved. I followed your [idea below](https://github.com/Automattic/wp-calypso/pull/72774#discussion_r1091035805) and simplified by using the already available `currentSitePlanSlug`.

I agree this can be simplified by refactoring `usePricingBreakpoint` (or even replacing it with a CSS solution) but considering the time constraints, it looks more like a post-launch effort.

File: client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx (hunk 6 above)
Comment (claudiucelfilip) [reply client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx]: Yes! Thanks for bringing it up! I now see there's a `currentSitePlanSlug` we can use.

File: client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx (hunk 7 above)
Comment (claudiucelfilip) [reply client/my-sites/plan-features-2023-grid/plan-comparison-grid.tsx]: Replaced with `featureGroup.slug`.

File: client/my-sites/plan-features-2023-grid/style.scss (hunk 8 above)
Comment (claudiucelfilip) [reply client/my-sites/plan-features-2023-grid/style.scss:888]: It's how I found it :D.
In the spirit of avoiding outside (of Plans comparison) changes as much as possible, I left it alone.
Considering there's more parallel work happening, I'd like to keep it as is to avoid unnecessary conflicts.

File: client/my-sites/plans/main.jsx (hunk 9 above)
Comment (claudiucelfilip) [reply client/my-sites/plans/main.jsx]: This version works too. Thanks!

**Review by dzver — APPROVED** [review]
It works but I have one question. Is it possible to make it so that in the smaller resolutions we have more plans in the visible part of the screen? The 2-3 split doesn't look well and there are no upgrade CTAs visible.

Screenshot 2023-01-31 at 11 09 14

Apart from that, I followed the test plan and it all worked well. I think there are also some questions by Christos that are not answered. Flagging as Approved to indicate that I find the change LGMT but would appreciate a response about the visibility of plans.

**Comment by claudiucelfilip** [comment]
@dzver Thanks for reviewing and approving!

> Is it possible to make it so that in the smaller resolutions we have more plans in the visible part of the screen? The 2-3 split doesn't look well and there are no upgrade CTAs visible.

There's a task to tackle this: https://github.com/Automattic/martech/issues/1459

> Flagging as Approved to indicate that I find the change LGMT but would appreciate a response about the visibility of plans.

I believe there's a task for this: https://github.com/Automattic/martech/issues/1450

**Comment by dzver** [comment]
Thanks!

**Comment by chriskmnds** [comment]
Thanks for addressing the feedback @claudiucelfilip ! 💯 😎
